### ISSUE TITLE
laser and grbl experiments

### ISSUE THREAD

**Issue by alpharesearch** [grbl/grbl#492]
Just like most comment here this is not an issue but I like to get some input and give some feedback. 
To play around a little bit I did mount a 2W blue laser+MOSFET to my ShapeOko+grbl and it is working. 
I did set it up for the variable spindle pin but didn't notice the min/max RPM settings at first (I guess I should have read all the comments). This cause some confusion because I was expecting 8 bit as default and not 0-1000. 
I did notice some dwelling when changing the laser intensity, but i guess MINIMUM_JUNCTION_SPEED should help with this (what value would be a good start for laser?). 
Now my main question would be are there any other important setting for laser setups to make?

**Comment by smartin015** [comment]
Did anything ever come of this? I'd really love to have this built in to grbl for my combo laser+spindle shapeoko.

**Comment by alpharesearch** [comment]
It is working for me now and I did submit a patch. I think @chamnit wants to wait for version 1.0 to add this feature? There are probably some other changes that need to be made and probably the patch will be obsolete. But the basic idea should still work with the new code. Just try my version and if you need help contact me.

**Comment by chamnit** [comment]
@alpharesearch : Yep. You are correct. The current working code, yet to be pushed,  is incompatible with your pull request. It'll be a nightmare to integrate it at this point and probably easier to write myself.

**Comment by alpharesearch** [comment]
@chamnit did you add a structure to be passed all the values from g code parser to all the motion function to the current working code yet? If so it would not be a big change at all to add the realtime spindle speed. We did discuss this at one point but I think you wanted to implement it your self.

**Comment by smartin015** [comment]
Uploaded your firmware version (grabbed it from your "laser" branch), and it works like a charm :) I did have to put in a line to import stdint in spindle_control.h to get the compiler to stop complaining.

Of course, I decided to do a speed test (500mm/min, 1000mm/min, and 1200mm/min), and got the following results:

![_mg_5150](https://cloud.githubusercontent.com/assets/607666/5640555/e2f4d1e0-95f4-11e4-82b8-07e1f3f8c104.jpg)
![_mg_5151](https://cloud.githubusercontent.com/assets/607666/5640556/e531d200-95f4-11e4-82af-1d67adac373b.jpg)
![_mg_5152](https://cloud.githubusercontent.com/assets/607666/5640557/e632033c-95f4-11e4-8fe9-a817cd3d2ec3.jpg)

What do you think? I can only think of a few causes to these discrepancies: poorly tightened belts, something being weird with the laser power module, or variable code timing in the ISR. The first doesn't describe the gaps in the middle of the lines, and the second would also show gaps at 1000 if that were the case. This makes me believe that the code is most likely the culprit.

**Comment by smartin015** [comment]
Also, disregard the black spot on the 1000 test - that's from another test.

**Comment by alpharesearch** [comment]
I think 3 things are causing the issues. There is a little line offset that is caused by the code, I could never figure out if there is a better place to change the power level change of the laser in the ISR. I can see this in the software simulator too. But my main thought here is that this is related somehow to the PWM output (but I don't know enough about this). However I did some text engraving on a piece of plastic (maybe 150mm/min) and this was not visible and I forgot about it because it worked for my needs. I'm never turning the laser off btw I only change the power level down to 0 (at this setting is looks like a normal dim laser pointer). If you want send me the 1200 gcode and I will run it through the simulator (or you can do that yourself, see my simulator patch). The other two factors here are the laser and driver and the medium. I did observe holes on some of my real world test with fast speeds on the standard firmware too, could small differences in the material maybe the culprit or the speed is changing? The longer startup holes maybe the driver not ramping up the power fast enough? What laser and driver do you have? How do you modulate the laser?

**Comment by smartin015** [comment]
You can find the 1200 gcode here: http://pastebin.com/t26Nd2kY

My laser is an ebay-acquired 2.5W blue laser (445nm) with a similarly ebay-acquired laser driver (http://www.ebay.com/itm/111103162677). I'm using an IRL520N mosfet in between the driver and the laser for modulation, with a 100k pull-down resistor. 

I'll hook a scope up to the spindle speed output and see if inconsistent PWM is the problem.

**Comment by smartin015** [comment]
Well, it looks like my signals are far from clean... 
![2015-01-07 19 39 06](https://cloud.githubusercontent.com/assets/607666/5656124/e958eb32-96a5-11e4-860a-e8e0dd350a41.jpg)

Could be I'm zoomed out a bit too far (1 div = 250ms), but those spikes outside of the main laser area tended to happen right when the head was reversing direction. Could possibly be wire getting noise from other parts of the machine... will investigate.

**Comment by rusirius76** [comment]
I'm working with this myself.  I just recently created a laser CNC project.  I'm using the Spark Concepts CNC xPro.  I updated the grbl to @alpharesearch  's version and set $40=1.

The problem I'm having still comes down to speed.  The S command is obviously the culprit.  If I generate gcode (using PEP 5+) which changes the Z axis for pixel darkness the code executes fine and speed seems to jive with whatever feedrate I request.  I just don't actually get anything since I'm not set up to encode the z to laser power.

If I instead use the same program to generate g-code using the S command to control power the gcode looks exactly the same except S in place of Z.  However when running this code the FASTEST feedrate (calculated by taking travel transit time and dividing out) is 425mm/minute.  I can set the feedrate to 600, 1200, 1500, 3000 it doesn't matter.  So obviously the S command is still causing a dwell or delay of some sort.  There has to be a solution to this.  So far my only option is to severely reduce laser power and run at a much slower 425mm/minute rate.  The images come out fantastic, it's just so slow compared to what I know it's capable of!

**Comment by alpharesearch** [comment]
@smartin015 I did run the file through the simulator and here are the results:
![martin](https://cloud.githubusercontent.com/assets/317055/5741858/0eff9ca6-9bdc-11e4-8938-f0b10a6738f4.png)
I'm still not sure regarding the offset, but maybe there is a better place in the ISR to change the PWM value? Like I said for my basic engraving (I did my name and address on the top and bottom of my chromebook) it worked just fine. If I remember right the error is not so bad at slow feed rate (I think I did 150mm/min). I think at this point we just need to wait for V1.0 and see what we can do from here. But regardless what Chamnit is doing it would be helpful to do some research to see how this can be implemented better. There was also a Kickstarter Laser CNC that is using GRBL and these guys probably did better implementation of the same thing... I think there was a promise to keep it open source on the Kickstarter https://www.kickstarter.com/projects/1537608281/lazerblade-the-affordable-laser-cutter-engraver. But this machine is a CoreXY, so there are other big changes. 

@rusirius76 I'm not sure, but if you look at the picture above the simulator thinks it is doing 1200. Because I only have a 3 watt laser I never tested anything faster 500 on my hardware and I never did take the time to make sure it really is going as set. I never did touch any of the planer code, maybe there is something that needs to be done at this level too? Maybe you want to run the 1200 file smartin015 did post and see if this is going 1200.

I think GRBL needs this kind of laser mode because it is so easy and inexpensive to slap a laser to any CNC now.

**Comment by chamnit** [comment]
@alpharesearch : Agreed. Grbl does need this. I'm getting ready to do a major push of about half the code I've been working on. At that point, I can revisit the laser support. The realtime spindle updating actually ties into spindle overrides as well, so I will need to work on this next or close to it.

**Comment by EliteEng** [comment]
@rusirius76 did you change your max travel speeds (max rate, mm/min) in the settings to above 500mm/min ?

**Comment by rusirius76** [comment]
@alpharesearch You should be able to push MUCH faster than that (when burning not cutting) wood with 3W.  I'm using an M140 diode that with my G2 lens maxes at about 2.3W output and I can burn a black line at well above 1300mm/minute.  You might be onto something with the planer though I don't know.  Every time I start trying to deep dive into the grbl code I start getting a headache.  :)  It's very interesting though, as I said I can take the same gcode file and swap the S's for Z's and it runs exactly the feedrate I specify... But as soon as I pop the S's in there it slows WAY down.  

@EliteEng absolutely, my x/y max rates are 12,000mm/minute and my z is 8,000mm/minute.  Well above what's being called for.  I don't remember my exact acceleration but it's very fast, not enough to be a factor.  Again I think this get's proven out by the fact that simply replacing the S's for Z's make it run at the requested speed.  Even up to 2500mm/minute (as fast as I bothered to test).  But like I said, as soon as those S's are popped in instead I drop to about 425mm/minute max no matter what the requested feed rate is.

**Comment by rusirius76** [comment]
@alpharesearch BTW... Looking at the gcode @smartin015 posted.  The biggest difference here is that he is making a full pass or line at a power setting... So he's sweeping the entire length...

Because my gcode is burning a "picture" in grayscale, I have a new gcode line for every pixel... So it's very small movements between them... .175mm at the moment to be exact...

For example, in his file a 40mm "line" would look like:

X40 S255

Mine however would look something like:
X.175 S50
X.35 S66
X.525 S78
X.7 S43
X.875 S66
X1.05 S168
X1.225 S213
X1.4 S200
X1.574 S155

........ etc... On up to X40

This is where I'm hitting this weird maximum limit of 425mm/minute...

However... If I pass the exact same code, but it looks like this...

X.175 Z4
X.35 Z4.1
X.525 Z4.15
X.7 Z4.03
X.875 Z4
X1.05 Z4.2
X1.225 Z4.26
X1.4 Z4.27
X1.574 Z4.4

........ etc... On up to X40

Then I get normal feed rates.  I've tested up to 2500mm/minute this way and it's fine, i'm sure much higher would be fine as well just didn't bother to test...  It's only the S command that slows things way down for me.

**Comment by rusirius76** [comment]
As a sample, here is a "grey bar" (just the first part) that I use for calibration to give you a better idea.

http://pastebin.com/xNTEN4VX

**Comment by alpharesearch** [comment]
@rusirius76 I did run your code through the simulator and it looks like 400mm/min max is the same result like your machine.
![xnten4vx](https://cloud.githubusercontent.com/assets/317055/5765262/20fccd44-9cc6-11e4-83a8-46e5ace7c8cc.png)
At least now we know for sure it is something in the code. I guess we need to compare the code execution between Z and S.
The processing for S starts at gcode.c at line 839 and 855, if the $40 laser bit is set to true it will only copy S or M3/M5 but not call the run function. At line 966 is calls mc_line for example and here the S value is passed on... but I only did pass the value down the chain (motion control, planer) to stepper.c; here I take the value and calculate the rmp at line 598 and I set the value at line 353. I'm not sure if calling the assignment through a function could slow everything down, but I don't think so because it it would run out of time and I think it would just not work at all.

**Comment by rusirius76** [comment]
@alpharesearch doesn't the busy flag in the ISR prevent that?  Bear with me, I'm a mediocre coder to begin with and have relatively limited experience with MCUs.  I play around a little here and there but that's about it.

It's funny, I was just digging through stepper.c when I saw your comment here.  I've looked through your code and on the surface I don't see an issue, the only possible thing I could come up with was call out to spindle_rpm_update and later the call to calculate_pwm_from_rpm.  I wonder as you, could either or both these calls be slowing things down enough to prevent the ISR from finishing before the next tick and the busy flag is just keeping things from going askew?

I honestly just don't see anything else that could be causing such a delay with the S command.

**Comment by EliteEng** [comment]
@alpharesearch the first place I would look is at calculate_pwm_from_rpm() being called in the prep buffer, this may be starving the ISR (segment_buffer) especially when there is very small moves.

The first thing I would do is move calculate_pwm_from_rpm() into gcode.c where mc_line is called, then have a uint8_t passed through the planner ( instead of a float ), this may help.

Another fix might be for gcode.c to only accept a uint8_t value then calculate_pwm_from_rpm() would not be needed.

Please note that these are only troubleshooting suggestions to try and find the cause of the slow down, and when the cause has been found then optimzation can occur.

**Comment by rusirius76** [comment]
Well, I just rewrote the calls and declarations to eliminate the calculate_pwm_from_rpm from the ISR.  I made the call from gcode.c when initially storing it in the struct.  From there everything was passed through down to the stepper.c level as a uint_8t.  I thought for sure this would do it, especially considering how often the ISR is firing off.  But nope, no change...  So back to the drawing board for that one.. The only other thing I can think of is the call to spindle_rpm_update, but I can't imagine that being it either... unless I screwed up or missed something in my patch, but it functions correctly so I can't imagine...

**Comment by EliteEng** [comment]
@rusirius76 since you have gone that far what happens is you eliminate calculate_pwm_from_rpm all together. if you make the S value in gcode.h a uint8_t instead of a float.

The only other thing I can see is that spindle_rpm_update is called with every tick of the ISR, this should have a "if(st.exec_block->spindle_speed_pwm != previous_pwm) then update". BUT this does not explain how @smartin015 is going to 1200mm/min with a constant pwm setting.

**Comment by rusirius76** [comment]
Another quick update... I eliminated spindle_rpm_update in the ISR and just set the register directly.  That had no effect.  

However, I'm not sure I totally understand the "dwell" time, but I just did a quick test... I set the default dwell time from 50 to 20 and reran my test... I've no increased to about 500mm/minute.  Still not great but much better, almost 25% better... So the first question is, what am I "really" doing by changing this?  What affect does this have on functionality?

Secondly while there could be other explanations, this kinda points to somehow having extra dwells taking place when sending the S command?  Despite the fact I can't seem to figure out from the code why...

**Comment by EliteEng** [comment]
@rusirius76 can you run your machine with a constant pwm setting like "G1 X100 F1200 S255" and confirm that you are getting speeds higher than 500mm/min with a constant PWM setting.

**Comment by rusirius76** [comment]
@EliteEng Sure, I'll do it as soon as I get home this evening.  Though I'm pretty confident it will.  I've cut paths before which would be exactly like this and never noticed any issue with speed.  But I'll definitely verify just to be sure!

**Comment by rusirius76** [comment]
Ok, so looking at the code, changing the dwell time doesn't make any sense.  It's only used as far as I can tell when can actual dwell (G04) is called.  So the increase in speed I saw must have come from the changes I made earlier to the calls.  I'm going to go back through and try to tidy up as much as possible and see if I can get it picked up any more.

**Comment by rusirius76** [comment]
ok, started trying to revise the code and ended up changing so much I just decided to fork off the master and start from scratch.  Just tidying up a bit (i.e. using the motion portions of the struct to determine if there is motion rather than setting a separate flag, etc. The rpm will get converted to PWM on input and carry through all the way.  I had a few minutes to work on it today and I'm close but still need to finish and test.  Won't get a chance tonight but hopefully tomorrow I can finish this up and let you guys know the results.

**Comment by EliteEng** [comment]
@rusirius76 another thing that might be causing problems is spindle_start() this is being called by gcode.c everytime a line of gcode is processed ( which resets the timer registers ). This may be causing delays.

I would move the contents of spindle_start() into spindle_init() just before the call to spindle_stop().
You would also need to change spindle_stop() from
"TCCRA_REGISTER &= ~(1<<COMB_BIT);"
to
"OCR_REGISTER = 0;"

**Comment by rusirius76** [comment]
@EliteEng spindle_start isn't being called every time, at least not that I see.  It's called on an M3, M4, M5 command and if an S command is received for a block that doesn't have motion assigned... If there is motion then only OCR_REGISTER is set.

I've cleaned up everything and reworked it slightly... But I seem to have hit a wall...  500mm/minute is the max I can get out of it.

I ran a standard file that was exactly the same as the normal code I'm running, except it had Z-axis moves instead of S and sure enough it was running at the requested 1200mm/minute.

So I also did as suggested and ran a file with S commands, but with a continuous pass instead of changing S for each pixel.

It too ran at the requested 1200mm/minute...  So it's not having the PWM enabled or anything like that... it's definitely the constant changing of the PWM.  I just can't grasp at this point what's causing it.

**Comment by rusirius76** [comment]
@EliteEng @alpharesearch @chamnit well, i figured out the problem, at least with my code, which largely borrowed from Markus' code just with heavy mods.  I'm just not sure what I'm missing.

In gcode.c during the execute phase, round about line 844 is the block of code (during set spindle speed) that uses an #ifdef for LASER_SPINDLE to bypass spindle_run if laser mode is enabled.  The problem is, it doesn't see LASER_SPINDLE as being defined. 

My code looks like this:
```
#ifdef LASER_SPINDLE
	    if ((gc_block.modal.motion != MOTION_MODE_NONE) && (axis_command == AXIS_COMMAND_MOTION_MODE) && (bit_isfalse(settings.flags,BITFLAG_LASER))) {
			  if (gc_state.modal.spindle != SPINDLE_DISABLE) { spindle_run(gc_state.modal.spindle, gc_state.spindle_speed); }
		  }
	  #else
	    // Update running spindle only if not in check mode and not already enabled.
	    if (gc_state.modal.spindle != SPINDLE_DISABLE) { spindle_run(gc_state.modal.spindle, gc_state.spindle_speed); }
    #endif
```

If I comment out the spindle_run call AFTER the #else then it works at normal speed... so for some reason it's not actually going into the block immediately after the #ifdef... Yet gcode.c includes system.h, which includes config.h which is where it's defined...  What am I missing here?

**Comment by alpharesearch** [comment]
@rusirius76 I think I saw some strange things like this compiling the simulator with config.h. Maybe it has something to do with some kind of circular header file detection in the compiler that is causing this... as a test if you take the whole '#ifdef' and everything after the else out but keep the if between '#ifdef' and '#else' it should work like expected (or just define LASER_SPINDLE right before manually). Btw Grblsim + debugger could show what would be executed.

**Comment by rusirius76** [comment]
yeah, it works as expected when I remove the #ifdef which for my purposes is fine, just would never be pulled that way, but I think Sonny wants to do his own anyway so no biggie.  

The only thing I have to figure out now is in gcode.c for setting spindle speed... I now have:

```
    if (gc_block.modal.motion == MOTION_MODE_NONE) { 
		spindle_run(gc_state.modal.spindle, gc_state.spindle_speed);
	}
```

But it's not working... an S command sent by itself on a line doesn't work, so it looks like gc_block.modal.motion doesn't get set or?  I dunno... Right now it's seems to be the last piece of the puzzle I need to solve... I suppose I could flip it and check for motion and default to running spindle_run... just seems sloppy...

**Comment by rusirius76** [comment]
Nevermind... I just fixed it by looking specifically for LINEAR motion mode, since g1 is the only time I'm worried about changing the spindle in real time...

```
if (gc_block.modal.motion != MOTION_MODE_LINEAR) { 
		spindle_run(gc_state.modal.spindle, gc_state.spindle_speed);
	}
```

Just tested and it looks like it's working perfectly!  I'll get the code pushed out to my fork in just a few for anyone who wants to have a look.  I'm getting 1200mm/minute with no problem at all now...  I'll do some burn testing and make sure everything else looks good.

**Comment by rusirius76** [comment]
@chamnit Sonny, BTW, looking at spindle_control.c since I was digging into the #ifdef statements... For variable spindle you are checking for CPU_MAP_ATMEGA328P... But in the CPU map hasn't it been changed to CPU_MAP_ATMEGA328P_TRADITIONAL and CPU_MAP_ATMEGA328P_NEW?  Shouldn't it be checking for one or both of these instead?

**Comment by chamnit** [comment]
@rusirius76 : Nope. Got rid of the TRADITIONAL and NEW suffixes in the last v0.9h push.

**Comment by alpharesearch** [comment]
I did change my code too (I will push it later), the only thing I'm doing in the ISR is setting the PWM. Everything else is moved to other areas that are not time critical. 

@smartin015 200mm/min vs 8000mm/min
![test1](https://cloud.githubusercontent.com/assets/317055/5818234/f1208556-a081-11e4-92cd-31c28b2d44f0.png)
@rusirius76 200mm/miv vs 1200mm/min
![test2](https://cloud.githubusercontent.com/assets/317055/5818236/f7452a90-a081-11e4-820f-8eeb5c13070f.png)

From the simulator it still looks like PWM is a little bit fuzzy at high speeds. But maybe this is just part of the simulator?

**Comment by rusirius76** [comment]
@alpharesearch Markus, I noticed this as well last night during testing.  I can GET the speeds now, but the change in PWM does seem to be slightly delayed at higher feed rates.  I haven't really had a chance to try to dig in and find out why yet.

Basically what I'm finding is anything up to about 1100 mm/minute is good and not really perceptible.. If you look VERY closely you can tell, but not much... On the other hand getting up faster results in much larger shifts...  

For example.  1100 mm/minute when burning a gray scale image (i.e. 1 gcode line per pixel) results in a pretty good quality image.  Running the same exact thing at say 1500mm/minute results in a very blurry image with almost no resolution at all.  The slower you run the higher quality the image (due to this delay causing pixels to overlap)

**Comment by alpharesearch** [comment]
I just merged the new edge with the laser patch and it looks like it still works. Somehow the new algorithm make the pattern look a lot more beautiful but at high speed is is still off. But it changed from a random pattern to a more predictable pattern. @chamnit does this ring a bell for you?   
@smartin015 file at 8000mm/min
![test3](https://cloud.githubusercontent.com/assets/317055/5826031/a7a47982-a0bd-11e4-92a9-80d89e24bb52.png)

**Comment by rusirius76** [comment]
@alpharesearch i've been digging into the code a bit and I'm having one of those moments...  You know where you have this vague foggy idea in your head of what the problem is, but when you try to really think about it it just vanishes until you stop thinking about it then it lingers back in?

I THINK the problem is one of timing... but I can't quite wrap my head around it... What I mean is, in this configuration the timer for the PWM output is never getting reset.  So at faster speeds I'm thinking it's ... well... now i'm going foggy again.. basically that it's getting out of sync and needs to be sync'd to the step pulse?  i'm thinking reset the timer when the step pulse timer is reset?  Does that make sense?

**Comment by rusirius76** [comment]
Ok, nevermind... i just realized why it was foggy.. cause I was being dumb... LOL... the step pulse timer IS what's generating the PWM... We are setting the counter...  hrmmmm... maybe we need to reset the counter?  let me dig in a bit more...

**Comment by rusirius76** [comment]
Ok, I'm not absolutely certain until I run more tests, but I think I just solved the problem... Maybe @alpharesearch can try it as well and see what the simulator shows.  What I did was to move my block of code that changed the pwm, i.e.:
```
		if (bit_istrue(settings.flags,BITFLAG_LASER)){
			if ((st.exec_block->spindle_direction != 0) && (OCR_REGISTER != st.exec_block->spindle_speed_pwm)) {
				OCR_REGISTER = st.exec_block->spindle_speed_pwm;
	        }
        }
```

UP in the ISR to the line immediately after it checks for the busy flag.  This way the PWM value is getting set to the PWM value for the last block (which could create an off by 1 error I guess, but since it's going to be even across the entire piece I don't see where this would be an issue) but more importantly (i think) this gets the counter set BEFORE the stepping pins are pulsed, step pulse timer is reset, and interrupts are re-enabled.  I'm still not sure exactly in my head why this matters, I can't really articulate it, but I had this nagging feeling that it was something to do with this.

I can't be sure without further tests, but I just ran a small test at 2000mm/minute and it looks good... whereas before I'm pretty sure it would have been a blurred mess.  Like I said, I need to test more with larger images where I can get a better idea, but I think it's solid.  But at this point I'm LONG over due for a good nights sleep!

**Comment by rusirius76** [comment]
The next step rolling around in my noggin is to use the st struct... my thinking is something along the lines of this, i'm just not sure exactly how I'm going to pull it off... it'll take some testing and experimenting...

Basically my thought is this....  We're updating the spindle speed (aka laser power) in real time... That's great, BUT... our motor speed isn't updating in real time.. it has acceleration and deceleration factored in...

What I'm thinking is, we use the PWM value as a "Target" value for the step we are on...  But we offset that value based on the acceleration or deceleration... 

For example, if a particular pixel (say a .15mm x .15mm space) is being moved over, and we want that pixel to be whatever darkness a "150" pwm value gives us....  As the speed of our motor ramps up, so too should the value of our PWM.  So that as the motors are moving across that .15 area slower, the value for the PWM is lower... Once up to full speed the PWM value will be exactly as requested...  When the speed of the motors drops below the feed rate on the other hand (i.e. deceleration) we need to do the same.  If our motors are running at half speed due to deceleration, then the PWM value should be reduced as well... overall this should give a much cleaner transition between "pixels"

What I have to wrap my head around now is how exactly to calculate that offset... I don't think it's as easy as a one to one ratio (i.e. if motors are running half speed, laser power should be half power).  Making a line burn at one speed and following it with another at half that feed rate isn't going to produce a line twice as dark.  Or I'm pretty sure it won't... So maybe this will have to be a variable that can be set for the offset... I dunno... It's pushing my knowledge of the grbl code to the edge.

**Comment by alpharesearch** [comment]
@rusirius76 yes we talked about changing the PWM depending on the acceleration and deceleration at one point here on github too. I would guess we would need to do these calculation in the planner.c? But my understanding is limited as well.
I did also do the test with the call after the busy flag, but I can not see any difference on the simulator output. Right now I'm testing using a flag bit to change it at the exact point... first I would like to be able to turn the PWM on at the exact moment when it should.

**Comment by rusirius76** [comment]
@alpharesearch Interesting... Like I said, I haven't had a chance to test, I just ran a small block and it seemed to show much better resolution than I would have expected at such a high feed rate.  

I've got a thought rolling around in my head, but honestly it's going to take a lot of thinking to implement I think... What it comes down to is this...

Determining a fixed number of cycles of the PWM... and then using stepper "movement" to run those cycles...  Ok, let me explain...

As I understand it, when stepper.c is running blocks, it's constantly outputting the step pulses at a fixed rate... Based on movement (and acceleration/deceleration, etc) these steps sometimes have movement, sometimes don't.  So only a portion of the step pulses actually "move" to the next step...  As the speed increases, the ratio of pulses that "move" versus pulses that just stay where they are changes.

What I'm trying to figure out, is if there is a way to "time" the PWM outputs with these "movements".  So let's say each "movement" produces 10 cycles on the PWM.  So each time a step pulse is fired off that actually MOVES the machine, 10 cycles are output on the PWM at the current duty cycle.  If the machine skips (i.e. doesn't move) for the next 30 step pulse cycles (because it's going slower) then the PWM doesn't have output either... As soon as the next step is made to move, then the PWM cycle is triggered again for that position...

Effectively this should decouple the laser power from the movement rate.  Which I think should increase accuracy and eliminate the issues we're seeing here.  As well as do what I mentioned earlier with taking into account acceleration or deceleration factors.  But now instead of having to "guess" at an offset, it's directly tied to the movement, so things should stay fixed and stable regardless of movement speeds, etc...

Does that make sense?

**Comment by chamnit** [comment]
@rusirius76 : You're way overthinking this. In the step segment generation, you have all the information you need to precalculate the PWM power based on speed and nominal power. Each step segment is executed as a constant velocity set of steps. You simply need to compute the PWM power based on the segment velocity relative to the programmed nominal speed and nominal laser power. That should be a very simple calculation.

**Comment by alpharesearch** [comment]
@rusirius76  please have a look at my latest edge push if you want; I introduced a flag bit to change the OCR It looks like it works a lot better now. The flag help with waiting for one extra cycle so that a change has more time. Plus there is a return to wait on stepping so this helps with the offset. From comparing it with the non real time way of doing it it has closer output behavior.

Regarding the changing the PWM depending on the acceleration I think we would need some kind of laser profile to be able to support different types of laser modules and maybe this could also take care of the non linear issues. I already moved the PWM value from the buffer to the segment struct... because all the acceleration stuff is done in the segment so we just need to change the pwn on ramp up/down to take the power down a bit... however there is already a workaround by just adding a white border around the picture before you export the g-code. By the time the laser reaches the picture the ramp up/down is already done and the engraving should have a constant speed in the engraving area.

**Comment by rusirius76** [comment]
@alpharesearch I'll have a look as soon as I get a chance. I'm not home most of this weekend unfortunately. I also created an update of mine off the 9h edge. I'll test both and see what the results are.

I also talked to Jeff, and with the pep5 it actually generates several white pixels (with pwm of 0) on each edge, so for that code at last acceleration shouldn't be an issue.

**Comment by skrutt** [comment]
I don't have much experience with lasercutters, but from reading this thread and what i know of pwm i don't think non-linearity will be an issue with scaling power vs acceleration. 
For example, a led is a very nonlinear device and therefore you cannot use an analog control method for them. But using pwm they work just the way you think they will, 10% pwm = 10% light.
Apply this to the laser and the same amount of energy should be delivered to the material if you scale power to acceleration. 
The observed non-linearity of a line not becoming twice as dark from twice the power i believe stems from the material not having a linear response to the laser. I guess you could test this by running a line at half speed and half power vs full speed full power, ie f500 s100 vs f1000 s200. If these lines have the same darkness scaling 1:1 vs acceleration should work very well, if not then things like small delays in switching the laser could be one reason or, maybe im wrong ;)
My 2 cent / Petter 
Date: Fri, 23 Jan 2015 12:32:40 -0800

**Comment by rusirius76** [comment]
@alpharesearch Markus, with your code in the simulator have you noticed any new "limit" in terms of top end speed?  With mine I'm now finding it maxes out at about 1700mm/minute when streaming the type of gcode I am (i.e. a gcode line for each and every pixel).  I haven't had time to test to see if it hits this same limit when using Z commands instead of S so I don't know if it's still a limitation somewhere or if i'm just hitting the limits of the baud rate, etc... But I know another guy who says he runs at 2200mm/minute+ so I don't think it's a limitation like that.. Unless he just doesn't realize he's being slowed...  I was wondering if you noticed any such limitation?  I just don't know rather it's worth trying to dig into the code again and look for anything else to tweak or if I should just accept it at this point. LOL...

**Comment by alpharesearch** [comment]
@rusirius76 as far as I remember @smartin015 gcode file was faster in the end - I think I had it at 8000mm/min. But it was a lot less gcode, so baud rate bottleneck may be possible? I need to run your file tomorrow again and see what I can find out. If you have a better file post it tonight and I will use the new file.

**Comment by rusirius76** [comment]
@alpharesearch Sure thing... Here's a much larger file... This one requests a 3000mm/minute feed rate, but again I was maxing at about 1700...

http://wikisend.com/download/431688/TreasureMap.nc

**Comment by alpharesearch** [comment]
@rusirius76 Yes I can see a very constant speed between 1600-1900mm/min (on the graph it has all the same purple color). I know if you have less changes it works faster. But I'm not sure where the bottle neck is right now. Could be streaming, gcode parsing, planer or to many ISR changes? To find it out can you generate the same file but without S changes and use Z for engraving? With with test we can let it run on the master and see how it compares.

**Comment by rusirius76** [comment]
@alpharesearch No problem... here's the same image with same options except using the Z-axis .005 to .255 movement....

http://wikisend.com/download/550190/1.rar

**Comment by alpharesearch** [comment]
@rusirius76 running the second file (I changed F1700 to F3000 in the file to match the first file) on the master branch shows exactly the same speed around 1700mm/min. I guess at this point we just him some other limit in grbl code or hardware. I'm not sure, but I think the speed is good for me and I don't need it faster right now.

**Comment by rusirius76** [comment]
was this with 9h or 9g?  I wonder if there is a difference?

**Comment by alpharesearch** [comment]
First 0.9g, but after your post just for grins and giggles I did test plain 9h too and it looks the same. Maybe 9h purple is a little lighter indicating a little fast speed... but nothing major.

**Comment by rusirius76** [comment]
@alpharesearch Well now I'm quite perplexed... Jeff (of PicEngraver) claims to be getting 2540mm/minute with no problems.  

He's using this fork of .8c... https://github.com/alsliahona/grbl/tree/LaserMode

Could it be that .8c was actually faster? He even tested two files, one at 1900mm/minute which took 7 minutes and 37 seconds to complete.  The same exact file at 2540mm/minute took 5 minutes and 52 seconds.  So he's definitely getting the feed rates MUCH higher than we are seeing.  

I just did a bunch of tests changing the junction deviation, step pulse time, and acceleration ticks... None of it made any difference in terms of the speed... So I just don't get it!

Perhaps @chamnit will chime in with any thoughts?

**Comment by EliteEng** [comment]
@alpharesearch what happens if you increase the BLOCK_BUFFER_SIZE in the simulator?

In your sim/config.h if you add the line
```#define BLOCK_BUFFER_SIZE 30```

Do you see a speed increase?

**Comment by chamnit** [comment]
@rusirius76 : I've read the code that that PicEngraver uses. It's fairly simple and uses v0.9g on a 328p. So, I don't think increasing the planner buffer would any effect. I haven't read @alpharesearch 's code thoroughly enough to tell you what the problem is.

**Comment by rusirius76** [comment]
@chamnit : Jeff told me he was using this code: https://github.com/alsliahona/grbl/tree/LaserMode which as far as I can tell is based on .8c no?

I've looked at it, and honestly it's very similar to what I'm doing in my own code.  The only real difference is that I'm overiding the S command and preventing it from calling spindle_run when it's received with a block that has a linear or arc movement... Instead it's passed down through the structs (as an int btw) to the stepper level where the OCR is set directly avoiding any calls to spindle_run, etc... The only time spindle_run is ever called is for an M3, M4, M5, or when an S is on a line with no motion or a seek.  With the code above, the one Jeff says he's using, the only real difference is that he's stripping the Z axis movements and setting the OCR in stepper.c based on it's value (when a flag is set).  Functionality wise I can't see any difference between his code and mine other than his is based on 8c and mine is based on 9h.

**Comment by rusirius76** [comment]
@chamnit : In reflection, I honestly don't think it can have anything to do with my code or @alpharesearch 's code.  Because we're seeing the exact same results when leaving the S command out completely and replacing it with a Z move.  This means that neither of the parts of code that we've wrote even come into play.  Which to me tells me it's an intrinsic problem in the 9g/9h code itself that for whatever reason doesn't effect 8c.

In trying to decode this a bit here's what we know...
1) Acceleration ticks / step pulse length / AMASS don't seem to affect it.  In fact, my machine can seek at 12,000mm/minute with no problems at all... So I can't see any problem with it not being able to execute the steps in time, etc...

2) A less "congested" gcode file has no problem... i.e. if I'm moving say 3mm for each gcode line it has no problems maintaining a much much faster speed.  The issue becomes more and more prevalent as gcode lines make smaller and smaller movements.  Currently I use .15mm pixel size for just about everything.  This means that for every .15mm of movement there is a gcode line...  Calculating this out means that at 1800mm/minute (which is the approximate cap I've found) grbl is processing 200 lines of gcode every second.  Conversely when requesting 2500mm/minute if it were actually able to do so would be 297 gcode lines per second.

3) The "sender" doesn't seem to be the problem since it can stream to his .8c based grbl at these rates without any problem.  Likewise obviously 115200baud isn't the limitation either since it obviously works ok with that version.

4) To me this would imply that the slowdown is happening somewhere either in the serial processing (i.e. gcode.c or protocol.c) or motion_control.c or lastly the planner.c.  So possibly the first place to look would be to see what changes were made to these between 8c and 9g and what if any of those changes may have resulted in a slow down.

**Comment by rusirius76** [comment]
I'm going to test something else when I get home tonight... Just a kinda wild thought... I'm using 1/16th microstepping... resulting in 80 steps per mm for my machine... I think Jeff is using a Shapeoko which I believe might be using 1/8th microstepping and getting 40 steps per mm.... If that's the case, then perhaps this is resulting in the bottleneck we're seeing?

**Comment by chamnit** [comment]
@rusirius76 : My apologies. I was thinking of different approach, not the Z-axis methodology that he uses. Like I mentioned before, I don't have a good feel of what you guys are exactly doing or what the problem is since I have not read your source codes in depth. Based on my experience with the code, you shouldn't be running to performance issues if installed correctly, but I could be wrong about this. Either there is something wrong with the code implementation, PWM doesn't like to be set in the ISR, or the grbl-simulator isn't providing an exact 100% representation of what the chip is doing (which may be likely). 

Like I stated before, one of the next things I need to tackle is overrides and spindle overrides need to actively update the spindle PWM realtime, much like the laser. When I tackle that, I'll install laser support. I'm sorry you are having so many issues, but if I run into the same problems, I'll know where to look.

**Comment by chamnit** [comment]
@rusirius76 : Had a glance at your code. Nothing outstanding. The only thing I would try is making some of the key variables volatile in the stepper.c code, like the step block PWM values. There might be a chance that the compiler isn't instructing the updates to the variables that the stepper ISR uses. By the way the code is designed, you shouldn't run into this problem, but you never know with compilers. They do some weird voodoo at times.

**Comment by chamnit** [comment]
@rusirius76 : And try disabling AMASS. It doesn't work as linearly as you'd expect.

**Comment by rusirius76** [comment]
@chamnit : First let me say that I really appreciate you taking time to think about this and step in to the conversation.  I know you've got a lot on your plate.

The thing is, I don't think it has anything to do with the PWM at all, because running the same exact code with a Z instead of S results in the same exact bottle neck...  i.e.

```
G1 X0 Y.15 S100
G1 X0 Y.30 S100
G1 X0 Y.45 S100
G1 X0 Y.70 S100

Results in a bottleneck of around 1800mm/minute...

BUT... So does
G1 X0 Y.15 Z100
G1 X0 Y.30 Z100
G1 X0 Y.45 Z100
G1 X0 Y.70 Z100

That's the perplexing part...  Yet opening up the moves (i.e. less gcode lines per distance) eliminates the problem...i.e.

G1 X0 Y1 S100
G1 X0 Y2 S100
G1 X0 Y3 S100
G1 X0 Y4 S100

has no problem hitting many thousands of mm/minute...
```

So it doesn't seem to have anything to do with the PWM portion....

**Comment by chamnit** [comment]
@rusirius76 : Ok. Thanks for the summary. This is quite a long thread and it's hard to digest what you've tried and what you haven't. 

So, I did a quick hand calculation on the distances you have and the speed you are running. Each block needs to be executed at a rate of every 5milliseconds. This is the CPU threshold of the 328p AVR chip (MEGA will not change this. same cycles). 5 milliseconds is roughly the time it takes Grbl to parse an incoming g-code command, plan the motion, and execute the steppers. This isn't probably what you want to hear, but it's about as good as you can get with an AVR and still have good motion control.

I'm surprised that v0.8 would run any faster, because I've done a lot of work increasing the end-to-end speed of the parsing-planning-steps code. It should be significantly faster than v0.8, but the demands of laser motion control could have found a relative inefficiency with the v0.9. So, at 2540mm/min and the same code, you'd have to complete parse-plan-step calculations in less than 3.4ms. That's a feasible possibility. Grbl v0.8 has to do overall less for each g-code block, while Grbl v0.9 is more optimized for high-speed over variable lengths of g-code motions per block. This would make a difference if you are hammering it with lots and lots of very short motion lengths.

**Comment by villamany** [comment]
Hi all, first many thanks for the great work.
Im trying some real world test, and for my perception seems to be very smooth movments up to about 4000mm/min, above this, the reported GRBL planing buffer go from 17 to unstable values (and start the mico-pauses between movments). Note if i made same test using a bluetooth HC05 device instead the USB cable i get very un-smooth movemets even below 1000mm/min.

For testing im using:
This laser engraver: http://3dpburner.blogspot.com
This Windows sender: https://github.com/villamany/3dpBurner-sender/blob/master/bin/Release/3dpBurnerSender.exe 
This GCODE file: https://drive.google.com/file/d/0B5Y1gEmSHMjMTTBRSEZxeUdqRTQ/view?usp=sharing

On the GCODE file you can replace the line 7 feedrate (s0 F1100.0) by desired speed.

However, for my setup i need no more than about 1100mm/min for plywood picture engraving.

**Comment by chamnit** [comment]
@villamany : This is a different problem altogether. This has to do with streaming and starving of the planner buffer. Laser motions are basically the worst-case scenario for planners (in general), because they involve very very short motions with relatively high speeds. Early on in Grbl's development, arc represented the same type of problem. We solved this in a number of ways, but there are limits to the CPU speed and the streaming interface. 

The problem you are having is involved with streaming. USB isn't a perfectly synchronous communication protocol. It sends and receives data in batches. For worst-case scenario g-codes, like laser motions, I would highly suggest using the character-counting streaming scheme, if you are already not. UGS and GrblPanel both do it already. If you are using a Bluetooth device, the streaming gets worse. The latency between Grbl and your computer becomes too great and Grbl will frequency starve. This is the reason why you are seeing it not perform as quickly as the USB.

You can try increasing the planner buffer size about 1 or 2 more blocks before Grbl (on Uno) runs out of memory. You'll know when you did by Grbl crashing. This should help a little. If you have an Arduino Mega, compile Grbl with something like 25-40 planner blocks. You should see a bit better performance.

**Comment by villamany** [comment]
Oh, i understand... Anyway im getting more than enought performance for my purposes.  

I tried some Windows senders... UGS crash on my computer many times, others dont use character count... so finally i create an app for my needs (using character count as reccomended on wiki) and getting a very good performance. Thank you!

**Comment by alpharesearch** [comment]
@villamany sometimes I use a Raspberry Pi + Wifi to remote control my machine. You can just upload the gcode file to the PI via wifi and start the job and turn your other PC off if you want. You can use plain terminal (ssh) with a steaming script or a graphical UI via VNC(or RDP) with UGS (or your own program) for example. This way the USB connection is via cable and you should get the best results. If you search on google you can find many different solution like this. I think there is a web interface for this too, but i never tried this. The first generation Pi was a little bit slow, but the new released Pi 2 should fix this. I will do some test in the next few week.

**Comment by villamany** [comment]
Nice! Waiting for your results. Im going to order a Pi now, it's the perfect excuse to get one :)

**Comment by scottrcarlson** [comment]
Make sure you grab the RPI 2 !

**Comment by rusirius76** [comment]
@chamnit @alpharesearch : Well, I've done a lot more testing and discovered that the problem isn't actually grbl at all.  It's the PicSender program.  It's using the call-response method of sending a single line of gcode at a time to grbl.  When I run the same files through UGC it has no problem at all hitting any requested feed rate.

I posted a video here doing a comparison.  http://youtu.be/blZhXLIXJ6Y

I'm in the process now of trying to urge PicEngraver to update their PicSender to use the advanced character counting method.  I'm not positive but it seems like they are going to be reluctant to do so since apparently it "works" for them.  Which I frankly don't understand.  I'm running on a pretty haus machine.

So anyway, I'll hope for the best, but in the mean time I've started writing my own sender program using the advanced character count method.  I'd just use UGC except it has a tendancy to crash with very large gcode files.  :(  So I guess I'll just do it myself!

**Comment by villamany** [comment]
@rusirius76 if you are using Windows try this one (click on raw for download). https://github.com/villamany/3dpBurner-sender/blob/master/bin/Release/3dpBurnerSender.exe

**Comment by rusirius76** [comment]
@villamany : Many thanks!  This quite honestly is perfect.  I was actually told that this program was VERY unstable with "small increment" gcode files.  I'm not sure why.  I just ran 6 different files, including the largest one I have which uses a .15mm pixel size and is 200mm x 250mm and it ran flawlessly.  Looks like I have a new sender program.  And open source at that.  Which frankly I love.  If I do find any stability issues I'll have a look at the code and see if I can fix it myself.  With just a quick glance though the code looks very good.

**Comment by villamany** [comment]
Nice! Hope you commit yours fix. 
I like the idea of simple program with only the basic functions: Put stock material, set the zero position where you want and start to work.

**Comment by PicEngraverToo** [comment]
@rusirius76 It keeps locking up when I try to run a gcode file that I have ran before on my Shapeoko with PicSender flawlessly. 

The one I ran in this video. https://www.youtube.com/watch?v=3naI5EsX89c

I tried every Mode setting and it gives me the same problem.

Greg, you do remember that we are running Andy's LaserMode grbl with Z0-Z255 PWM commands and variable feedrate on every line of code, right? The instability issue I mentioned seems to be a grbl or a variable feedrate compatibility issue. Have you tested our PicEngrave Pro 5 Feed Rate Change option with that sender? Did you ever even try Andy's LaserMode grbl as I asked you to do so many times with PicSender? 

We have no problems here with our streamer. Good luck with your testing!

**Comment by alpharesearch** [comment]
I received a tracking number but I don't have it yet... if you want WIFI 
you need to gt a USB dongle. This is the WIFI USB dongle I use: 
Edimax-EW-7811 for $8.99 via Prime at Amazon for example. The good thing 
about this one is that the drivers are part of the base system.

**Comment by Dikabro** [comment]
Hi guys, check out this software for laser engraving: http://happycoders.com/release/laser/ :)

**Comment by rusirius76** [comment]
@PicEngraverToo  I haven't tried it with Z commands, but I did test it last night with variable feedrate.  I didn't have any problems so I'm not sure.  I also ran the largest file I've done to date.  A 351mm x 258mm @ 2500mm/minute and had no problems at all.

I haven't tried Andy's grbl but I did do a very extensive dive into his code when you mentioned not having any problems with faster feed rates with his code.  His code is essentially exactly the same as mine (and @alpharesearch 's as well for that matter).  The principle is exactly the same, pass the PWM value down through to the stepper.c and then change the PWM comparator on the fly there as the block is processed.  Andy's code is based on .8c and I dunno, call me an eccentric but I don't like the idea of running such an old release of software when so many other improvements have been made.  

If his code worked great and 9g or 9h absolutely refused to work then I'd certainly fall back.  But when I can get perfectly functioning results with other senders and it's only PicSender that's the limitation I can't convince myself it's worth going back to this old version just so I can continue to use PicSender.  Especially since the 3dpburnersender is open source.  I can adjust or modify it any way I see fit.  And if I do run into any problems with lockups, etc I can debug them and fix them.  Though honestly as of right now I'm running my 9th stream with it and it's been flawless so far.

It's not that I don't like your software... I do... But... You have to try to understand from my point of view...

Think of it like this...  If I decide to paint my house...  I go out and buy a brand new professional grade paint sprayer that has all the bells and whistles.  Then I go and buy a bucket of paint... It's great paint, but when I load it up in the sprayer I find that I'm getting clogs constantly... The paint is a great brand, and I like the manufacturer.  I try to make it work... I do all sorts of testing, but what I find is that a bunch of other paints I try spray just perfectly from my brand new sprayer... It's only this paint that doesn't work... and after testing I find that it's because of a certain non-recommended method they use in their manufacturing.

So I contact the manufacturer... Hoping that with all the testing and proof I've gathered they'll tell me that since it's only a benefit to the entire paint community, they'll work on implementing this new method that's the recommended way of doing things... The way the other paint manufacturers are doing.  I mean, it won't cost them anything other than a little time, and again, it just makes their paint that much better than it already is.

Instead the paint manufacturer tells me that they are using Wagner paint sprayers... and even though they're older technology they work very well.  They suggest I throw away my new professional grade paint sprayer with all the bells and whistles and instead buy an old Wagner sprayer because it will work fine with their paint.

Now be honest with yourself... If you were in this situation... Would you get rid of your new paint sprayer and get a Wagner?  Or would you just move on and use a different paint manufacturer?  Especially if you found that manufacturer allowed you to formulate and customize your paint in any way you'd like?

**Comment by Dikabro** [comment]
What microstepping do you using on stepper drivers for picture engraving? 
I having some problems with pictures engraving and trying to figure out what's wrong... I'm thinking that maybe microstepping is causing problem. I attached a picture of what I get when engraving grayscale pictures. Please, maybe anyone could suggest for me something?
![screenshot 2015-02-08 19 04 53](https://cloud.githubusercontent.com/assets/10909354/6097411/b2e24e84-afc5-11e4-95ac-9a26a0d92a24.png)


My system:
1. Analog laser driver http://www.ebay.com/itm/3A-analog-laser-diode-driver-with-thermal-protection-for-445nm-638nm-520nm-diode-/161353680119?ssPageName=ADME:X:AAQ:US:1123
2. Arduino Nano with 0,8c laser mode grbl.
3. NEMA17 stepper motors
4. A4988 Polou stepper motor drivers.
5. 400 mW red laser

**Comment by villamany** [comment]
Im using 16microstep. What is the problem? Inverted image?



Enviado desde dispositivo movil.

-------- Mensaje original --------
De: Dikabro <notifications@github.com>
Fecha:2015/02/08  18:07  (GMT+01:00)
Para: grbl/grbl <grbl@noreply.github.com>
Cc: villamany <villamany@msn.com>
Asunto: Re: [grbl] laser and grbl experiments (#492)

What microstepping do you using on stepper drivers for picture engraving?
I having some problems with pictures engraving and trying to figure out what's wrong... I'm thinking that maybe microstepping is causing problem. I attached a picture of what I get when engraving grayscale pictures. Please, maybe anyone could suggest for me something?
![screenshot 2015-02-08 19 04 53](https://cloud.githubusercontent.com/assets/10909354/6097411/b2e24e84-afc5-11e4-95ac-9a26a0d92a24.png)


My system:
1. Analog laser driver http://www.ebay.com/itm/3A-analog-laser-diode-driver-with-thermal-protection-for-445nm-638nm-520nm-diode-/161353680119?ssPageName=ADME:X:AAQ:US:1123
2. Arduino Nano with 0,8c laser mode grbl.
3. NEMA17 stepper motors
4. A4988 Polou stepper motor drivers.
5. 400 mW red laser




---

**Comment by Dikabro** [comment]
The problem is that in my engraved pictures exist only black and white "collors", gray doesn't exist and I don't understand why..

**Comment by villamany** [comment]
I dont know abot the 0.8 PWM  capabilites (i only tested 0.9), maybe someone more experienced can help you. But you can post the GCODE file for discard a file generation issue.

**Comment by rusirius76** [comment]
@Dikabro : What medium is that?  In the picture it almost looks like a fiberglass or something like that? Could just be the pic.

Some materials have a very nice grayscale capability... Others are very touchy requiring fast feed rates.  Poplar for example I find pretty easy to get gray scale on, though faster feed rates open up the range.  On the other hand, some white woods I've found (don't know exactly what they were) were just almost a 1-bit medium where it was very difficult to get any sort of variation in color.  Basically once it burned it just burned.

It could be the medium or it could just be that you need to get your feed rate up faster?

**Comment by PicEngraverToo** [comment]
@rusirius76  Try testing PEP5 & PicSender and consider your laser diode as paint and a sprayer and lower it down and aim it at your material instead of waving it in the air and complaining that it does not perform like you want it to. :-D

One customer's criticism does not warrant a change. ;-)

**Comment by alpharesearch** [comment]
@rusirius76 have a look at my imagetolaser project... maybe you could help with some real world data... I could make a QT GUI and we could make it available for windows too but I need to refine it but I'm not sure about the exact requirements. I only made it (or better forked it) for a quick tool to get the simulator tested because I didn't wanted to fork over any money ...https://github.com/alpharesearch/imagetolaser
But I think it should work just fine and it is foss, we could combine this with a sender too. If we find a foss sender it should be easy to integrate or merge the two.

**Comment by kfoltman** [comment]
Speaking of frontends and QT GUIs, I'm also working on something related - a PCB CAM and a Python GRBL sender/frontend - https://github.com/kfoltman/wharrgrbl - the sender part is not functional yet, but pretty close. The character-counting sender is already implemented, and it seems to work.

So, chances are, if we combine the efforts, we might end up with something that can do PCBs, laser engraving of pictures and maybe even laser exposing (think UV soldermask) and have it drive GRBL directly. This gives some extra possibilities due to integration, for example, auto-Z-probing or redoing the underetched areas on request.

**Comment by Dikabro** [comment]
@rusirius76 I dont know on what kind of material I engraving, but it is a little bit glossy, maybe it's lacquered. I will try to engrave on different material and will post results here. :) 

Do you using  0,9 grbl version on older 0,8c lasermode edition? 
If you are using 0,9 grbl version then maybe you could send your code for me in private? :)
But don't send only hex file, because in my system there is  Arduino NANO and I need to change pinouts..:)

Last time when I was flashed 0,9g master version on my NANO - my homing function doesn't worked correct like in 0,8c...

**Comment by rusirius76** [comment]
@alpharesearch : Very cool... It's funny, I actually just started some code this weekend that will be FOSS.  I'm working in visual studio though for a windows platform.  .NET already includes a lot of great features for image manipulation.  It literally took me about 10 minutes to have some functioning "guts" that would allow you to open any image (.gif, .jpg, .bmp, .wmf, .png, etc), converts it to grayscale (averages the Red, Green, Blue color components of each pixel) and then rewrites those values as that average.  Then can do the same to generate gcode based on the values and locations of those pixels.

There's a lot to do in terms of refining it, making it slick and intuitive, and adding lots of features, but it really shouldn't take much development time.  

I love what I'm seeing here though... This is exactly what this community needs.  A lot of options and tools to make these sorts of projects as accessible as possible.

I'd be glad to test out the imagetolaser and throw out any ideas that come along!

@kfoltman : This is great! I'm actually a bit of an electronics nut myself and am awaiting the arrival of some dry UV film.  I wanted to do exactly that, explore the PCB making abilities.

@Dikabro : Yeah, definitely change material.  Poplar and Birch work really well for me.  I also make sure to sand the surface before etching.  Usually 320 grit minimum... Often times polished up with some 600.

I use .9h.  Both @alpharesearch and I have versions of it that have been modified for "real-time" spindle control (i.e. laser power adjustments via S commands don't slow things down.)  I'd recommend using his though because to be honest I don't remember the last time I pushed mine and I've been working on it quite a bit to clean it up.

**Comment by villamany** [comment]
@rusirius76. Yesterday i start a similar project! For the moment i have GUI about finished (very simple). I can convert any image to grayscale, adjust the brightness, contrast and gamma by friendly sliders and then generate a bitmap of the correct size based on desired resolution and desired output image size (in mm) I only need to get each pixel and convert to gcode..maybe tomorrow i can start to generate the first gcodes. Im planing also to add some type of profiles for minimize the no linear output power.

**Comment by rusirius76** [comment]
@villamany : Awesome!  I'm looking forward to seeing it when it's flushed out!

**Comment by rusirius76** [comment]
Ok, I've worked a little bit in my free time to add some image manipulations.  Currently it will flip, rotate, invert (make negative for mirror etching, etc), etc... I just finished the dithering routine... I set it up so that other methods can be used (it uses a dither matrix that can be defined based on how you want the error distribution to be applied), but honestly I think it's probably best (and certainly more simple from an end user perspective) to just use Floyd-Steinberg.  In my experience it generally produces the best results anyway.

The entire concept with this is to make it as easy as possible for an end user.  They don't have to worry about image size, etc.. Just tell it the pixel size you want, the height and width you want the finished gcode to burn, etc...

**Comment by villamany** [comment]
i'm starting some test for image to gcode conversion, could someone confirm me if i am using the correct logical for minimize the data stream between PC-GRBL? 

-Remove all white spaces from lines 
-Send the modal G1 command only at begin instead on each line
-Add only the CR as line end, instead CR+LF 
-Send only the  X, Y and S values if they are diferent since last update
-Using the format 0.### for coordinates

**Comment by rusirius76** [comment]
@villamany : Sounds right to me!

**Comment by rusirius76** [comment]
@villamany : also, you have to be careful, so you don't overrun any limits.. but I plan on incorporating a user selectable option for a "buffer" zone around the image.  So for example, 5mm of "blank space" around the sides of movement... So that the machine is "up to speed" by the time it actually starts burning pixels... No acceleration during burning...

**Comment by alpharesearch** [comment]
@kfoltman sounds like a very interesting project, I think I need to start learning KiCad again... I once made I little PC keyboard with it: http://alpharesearch.de/?page_id=53 but at that time I use batchPCB. 

@rusirius76 I like .net and C#, but if you want to support not just Windows make sure to make it mono compatible.

Regarding the Raspberry Pi 2 it is a lot faster now, I did use the latest version of UGS via VNC and it starts fine and the web page pendant feature was accessible from my other PC. To copy files I use scp (mc under linux or winscp under windows) I think it is more usable now and I'm happy with it.

If you get the latest raspbian image from raspberry pi web page you have Oracle Java installed, but you need to get the rxtx library for serial communication.

```sudo apt-get install librxtx-java```

now you can start UGC like this:

```java -Djava.library.path=/usr/lib/jni -cp /usr/share/java/RXTXcomm.jar -jar -Xmx256m $rootdir/UniversalGcodeSender*.jar```

**Comment by villamany** [comment]
Here are my first working results for image to gcode conversion:

https://github.com/villamany/3dpBurner-Image2Gcode

.exe file: https://github.com/villamany/3dpBurner-Image2Gcode/blob/master/bin/Release/3dpBurner%20Image2Gcode.exe

**Comment by Dikabro** [comment]
Very nice work!!! Simple and good.;)

**Comment by rusirius76** [comment]
@villamany : Looking good!  I'm actually coding mine in VB.Net.  For what I do in my job I end up having to use it a lot more so development for me just comes much faster using it.  And with .NET there isn't really much difference anyway!  Dithering is an important aspect though, Sometimes images just come out "better" with dithering I've found, and for mirrors or other essentially 1-bit mediums it's just a must.  Here's the code I use for it... Should make your life a little easier...

```
        Dim bm As New Bitmap(pbRender.Image)
        Dim X As Integer
        Dim Y As Integer
        Dim clr As Integer
        Dim DitherTable(0, 0) As Byte
        Dim dErrors(0, 0) As Double
        Dim errorVal As Integer
        Dim dDivisor As Integer
        Dim xLeft As Integer, xRight As Integer, yDown As Integer
        Dim QuickX As Integer, QuickY As Integer

        'The core code is run against a "dither table" which contains distribution of the errors.  This way
        'other dither methods can be implemented later by just supplying a different table.
        'First, prepare a Floyd-Steinberg dither table
        ReDim DitherTable(3, 2)
        DitherTable(2, 0) = 7
        DitherTable(0, 1) = 3
        DitherTable(1, 1) = 5
        DitherTable(2, 1) = 1
        dDivisor = 16

        'Second, mark the size of the array in the left, right, and down directions
        xLeft = 0
        xRight = 2
        yDown = 1

        'Redim a dithering table the same size as the image.
        'We could just use a 2 high array instead of the entire image and reset values each time we
        'move to another line, but I think this should be faster, and memory shouldn't be a problem
        'for modern machines.
        ReDim dErrors(pbRender.Image.Width, pbRender.Image.Height)

        'Now loop through the image, calculating errors as we go
        For X = 0 To bm.Width - 1
            frmStatus.pbStatus.Value = X / (bm.Width - 1) * 100
            For Y = 0 To bm.Height - 1
                'When the image is loaded it's converted to grayscale, so R, G, and B are all the
                'same value for a given pixel, so we just use G.
                'Set clr to the original value + any error stored in the error table.
                clr = (CInt(bm.GetPixel(X, Y).G)) + dErrors(X, Y)

                'Now let's figure out if it should be black or white.
                If clr >= 127 Then
                    'We're setting the pixel black, so the error is the difference between original pixel
                    '(plus error applied) and 255.
                    errorVal = clr - 255
                    'now set the pixel black
                    bm.SetPixel(X, Y, Color.FromArgb(255, 255, 255))
                Else
                    'Setting it white.  Error is just the original pixel value + error applied
                    errorVal = clr
                    bm.SetPixel(X, Y, Color.FromArgb(0, 0, 0))
                End If

                'If there is an error, spread it according to the dither table
                If errorVal <> 0 Then

                    'Spread across the relevant pixels according to the dither table formula
                    For i = xLeft To xRight
                        For j = 0 To yDown

                            'ignore already processed pixels
                            If (j = 0) And (i <= 1) Then Continue For

                            'ignore pixels that have a zero in the dither table
                            If DitherTable(i, j) = 0 Then Continue For

                            QuickX = X + i
                            QuickY = Y + j

                            'Next, ignore target pixels that are off the image boundary
                            If QuickX < 0 Then Continue For
                            If QuickX > bm.Width - 1 Then Continue For
                            If QuickY > bm.Height - 1 Then Continue For

                            'Now apply the error here.
                            dErrors(QuickX, QuickY) += (errorVal * (CSng(DitherTable(i, j)) / dDivisor))
                        Next j
                    Next i
                End If
            Next Y
        Next X
        'We've swept the whole image, so let's put it back in the picturebox.
        pbRender.Image = bm
```

**Comment by villamany** [comment]
Nice! I will try to implement it.

**Comment by chamnit** [comment]
Ok fellas. This topic has veered off course and has gotten out of hand. I'm closing this issue. Should you want to open another one on the specific topic of laser GUIs for Grbl, go ahead.

**Comment by rusirius76** [comment]
I was actually thinking the same thing... I went ahead and created a forum for everyone here to discuss pretty much anything to do with laser engraving.  It's located at www.engraving-lasers.com.  Lot of work to do yet in setting it up, but it's there at least and it'll let us discuss this stuff somewhere more appropriate.  Sorry for running over your issue thread @chamnit.

@alpharesearch 
@villamany 
@Dikabro 
@EliteEng

**Comment by zesava** [comment]
Hi everyone. I will leave here my five coins :) I have builded 3dpburner by villamany and was confused how fast i may tuning laser power/feed rate for different materials. So i wrote my own app. 
![screen](https://cloud.githubusercontent.com/assets/15175423/10561357/cec0d188-752d-11e5-9837-5855c69d6aec.PNG)

with it you can adjust parameter during engraving. App was tested with  AlphaResearch version at Feb/04/2015 and with regular grbl v0.9i for Z head mode. Hope someone found this usefull.
https://drive.google.com/open?id=0B_ZjDNHJ9krcZC12bDRfZjFkc0U

**Comment by Dikabro** [comment]
Can't import 8bit picture...

**Comment by zesava** [comment]
App accepts only  8bit *.BMP.

**Comment by Dikabro** [comment]
Could you make 16 bit modification?

**Comment by zesava** [comment]
if you look at grbl source code(config.h), you will find next line


*// Use by the variable spindle output only. These parameters set the
maximum and minimum spindle speed// "S" g-code values to correspond to the
maximum and minimum pin voltages. There are 256 discrete and // equally
divided voltage bins between the maximum and minimum spindle speeds.*
so there is no sense to make 16 bit image resolution, coz 16 bit means
65,535 colors of gray, but we have only 256(8bit) output. So this
modification will have no influence for the engraving.


2015-10-19 23:45 GMT+03:00 Dikabro <notifications@github.com>:

> Could you make 16 bit modification?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grbl/grbl/issues/492#issuecomment-149342064>.
>



-- 
Live Free Or Die Hard

**Comment by Dikabro** [comment]
I understand that. But the problem is that I can't open any image on your software. It doesn't allows for some reason.

**Comment by zesava** [comment]
http://lmgtfy.com/?q=8+bit+grayscale+bmp

2015-10-22 20:56 GMT+03:00 Dikabro <notifications@github.com>:

> I understand that. But the problem is that I can't open any image on your
> software. It doesn't allows for some reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grbl/grbl/issues/492#issuecomment-150305840>.
>



-- 
Live Free Or Die Hard

**Comment by StuartB4** [comment]
@Dikabro. 

You have to click the GRBL button and connect to the COM port first. Then click the GCODE button and the Open image button is available.
